### ISSUE TITLE
Orphaned extension host instances after running tests

### ISSUE THREAD

**Issue by StephenWeatherford** [microsoft/vscode#81397]
Issue Type: <b>Bug</b>

I don't have repro steps right now. If this is an unknown issue and you can't repro, let me know, I'll find something.  

Very frequently when running suites, if the suites don't end successfully, the extension host instance doesn't not get closed.  This is a regression from stable version.

VS Code version: Code - Insiders 1.39.0-insider (8b6e0efc0a98639085fabeb5226a516785424796, 2019-09-19T03:15:21.720Z)
OS version: Windows_NT x64 10.0.18362

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Xeon(R) W-2133 CPU @ 3.60GHz (12 x 3600)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off<br>surface_control: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.73GB (12.18GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (9)</summary>

Extension|Author (truncated)|Version
---|---|---
gitlens|eam|10.0.1
vscode-cosmosdb|ms-|0.11.0
vscode-docker|ms-|0.8.1
remote-ssh|ms-|0.46.1
remote-ssh-edit|ms-|0.46.1
remote-ssh-explorer|ms-|0.46.1
azure-account|ms-|0.8.4
csharp|ms-|1.21.3
vscode-typescript-tslint-plugin|ms-|1.2.2

</details>

**Comment by StephenWeatherford** [comment]
BTW, this is when debugging a vscode extension with vscode.

**Comment by StephenWeatherford** [comment]
Also, if I'm on a breakpoint, pressing F5 frequently shows the error "There is already a debug configuration "Launch Extension" running."

**Comment by chrmarti** [comment]
Is this only with tests?

**Comment by isidorn** [comment]
@StephenWeatherford for the F5 showing "there is already a..." this is duplicate of https://github.com/microsoft/vscode/issues/81373 and will be fixed in tomorrows vscode insiders

As for the vscode window not closing. I can not reproduce this. Here are my steps:
1. I have a vscode starter extension sample
2. I run the extension tests. Window opens and nicely closes

Let me know if you are doing anything differently (gif attached)

We heard other reports of this, so there is some issue here however it would be great if we could find some steps to track this down.

Thanks for using vscode insiders!

![tests](https://user-images.githubusercontent.com/1926584/65609187-a4e12480-dfaf-11e9-8237-de1087efe2ba.gif)

**Comment by StephenWeatherford** [comment]
Happening quite a lot with my extension, although inconsistently.  Try this:

1) git clone https://github.com/microsoft/vscode-azurearmtools.git
2) cd vscode-azurearmtools
3) npm i
4) Load folder in vscode and F5 with configuration "Launch Extension"
5) Open a file, change language to "Azure Resource Manager Template"
6) Go back to the original vscode instance and press the stop button
!) Frequently for me, the extension host instance doesn't close

Try it multiple times, maybe with different lengths of time before stopping, hopefully it will repro for you.

**Comment by StephenWeatherford** [comment]
Any more info?  This is extremely annoying, I keep ending up with 5 or 6 instances that I have to close.

**Comment by StephenWeatherford** [comment]
![image](https://user-images.githubusercontent.com/6913354/65914996-5228b380-e387-11e9-92f2-54b8eb28bbd1.png)

**Comment by alexandrudima** [comment]
I have also tried on Windows with the steps in https://github.com/microsoft/vscode/issues/81397#issuecomment-535117239 but the window goes down every time I press the stop button

**Comment by alexandrudima** [comment]
@isidorn I'm just pressing F5 and clicking the Stop button:

![TO_UPLOAD](https://user-images.githubusercontent.com/5047891/65968456-79ed4980-e463-11e9-9bbc-dd096d6b5a1c.gif)

**Comment by StephenWeatherford** [comment]
Nope, debugger is attached.
![image](https://user-images.githubusercontent.com/6913354/65994565-7ac4b200-e448-11e9-8b61-29072e976e47.png)

**Comment by isidorn** [comment]
I can reproduce with @alexandrudima steps.
So start debugging, as soon as the window is opened click stop. The window is left orphaned, however in that case the Debugger is not attached. I can also reproduce this on Stable, so I am pretty sure this is not the issue that @StephenWeatherford is hitting. I will now try to reproduce using exact @StephenWeatherford steps both on mac and on windows..

**Comment by alexandrudima** [comment]
@StephenWeatherford Is there anything else we are missing to reproduce on our side? I haven't seen the log statement "Initializing dotnet acquire..." when I tried things out, maybe that has something to do with being able to repro ?

**Comment by StephenWeatherford** [comment]
I don't know what changed, but I haven't been able to repro since the update a couple days ago.  Thanks for looking into it.

**Comment by StephenWeatherford** [comment]
It's back with Version: 1.40.0-insider (user setup)

**Comment by StephenWeatherford** [comment]
Doesn't seem to be reproing in the latest stable, just insiders.

**Comment by chrmarti** [comment]
Could this be caused by the Electron 6 update? /cc @bpasero

**Comment by isidorn** [comment]
@chrmarti good point. Yes, the timing works perfectly, since @StephenWeatherford mentinoed it stopped ocurring when we reverted back to previous electron, and now reprots it happens again when we updated again to 6.

**Comment by StephenWeatherford** [comment]
@deepak1556 Well, now I'm seeing it in stable, too.

**Comment by StephenWeatherford** [comment]
Happens a lot for me when I'm debugging suites, but don't get consistent results when trying to create a repro. Try this:

1. git clone https://github.com/microsoft/vscode-azurearmtools.git
1. cd vscode-azurearmtools
1. git checkout saw/vscode-81397
1. npm i
1. Load folder in vscode and F5 with configuration "Launch Tests"
1. Open test\UserFunctions.test.ts
1. Set breakpoint at line 129
1. F5
1. At the breakpoint, press the stop button
!) Frequently for me, the extension host instance doesn't close

**Comment by bpasero** [comment]
Adding some people to try to reproduce.

**Comment by bpasero** [comment]
I cannot hit the breakpoint and I see no hang. I have no idea what this is.

**Comment by alexandrudima** [comment]
@StephenWeatherford I see the following when launching the tests. The breakpoint is also not hit for me:

![image](https://user-images.githubusercontent.com/5047891/68786479-2401e900-0640-11ea-81e5-9c3b1ec85f21.png)

Maybe there is some setup step which we are missing?

**Comment by StephenWeatherford** [comment]
@alexandrudima Sorry, launch.json had wrong setting, please try again.

**Comment by bpasero** [comment]
I can reproduce but not by just having  1 breakpoint and stopping. For me I need to have a breakpoint above the line, stop there, continue and then stop. It also seems to trigger if I add another breakpoint while being stopped. 

Maybe @weinand or @isidorn could comment how exactly the click on the stop button wires through to the window closing.

**Comment by isidorn** [comment]
@StephenWeatherford thanks for updating the `launch.json`, now I can run the tests. However I can not repro this on my osX. My Win machine is very slow, so I plan to try this once I get a new laptop (hopefully in a couple of weeks).

@bpasero Clicking the stop button will send out the `disconnect` request to the node adapter. The node adapter will recieve this and take down the EH process. The workbench detects the EH going down and takes down the window.
That is my understanding of the flow. @weinand @roblourens correct me if I am wrong

**Comment by bpasero** [comment]
For me:

https://github.com/microsoft/vscode/blob/2305f2187475d860f5f70c8d3bff548f6ce08929/src/vs/workbench/services/extensions/electron-browser/extensionService.ts#L573

Is not getting called.

**Comment by bpasero** [comment]
More specifically, this one seems to be the trigger normally:

https://github.com/microsoft/vscode/blob/2305f2187475d860f5f70c8d3bff548f6ce08929/src/vs/workbench/services/extensions/electron-browser/extensionHost.ts#L216

**Comment by bpasero** [comment]
@isidorn @weinand possibly the way you take down the process does not work anymore when a breakpoint is hit?

**Comment by alexandrudima** [comment]
@bpasero We looked at this here together with @isidorn on my machine. We have noticed that the extension host is still running, and is still paused at the breakpoint. We have connected to the extension host process from Chrome via its inspector and once we pressed continue, the extension host process ran to completion and the window closed.

So the issue is that pressing "stop" (the red box) is leaving the extension host in a paused state at the breakpoint. @isidorn has suggested that this should be investigated by @roblourens as it is a node-debug2 issue.

**Comment by isidorn** [comment]
@bpasero yes you are correct.
After investigating with @alexandrudima we have found out that the Extenino Host is not being shut down. It remains in the stopped state, since we connected to it via Chrome and verified it is still on the stopped breakpoint.
So the issue is that on Windows with newest electron the node debugger can not shut down the extenion host if it is stopped.
As a workaround the debugger could send the continue before shutting it down

**Comment by roblourens** [comment]
I can repro this, but in my case it is using a random port, not 9229. It should be using a random port unless 9229 has been specified, do you know why that would happen @isidorn?

Also, I can't repro this in the exploration build, so it could be an electron issue too.

**Comment by roblourens** [comment]
Also, the node debug adapter does not try to kill the extensionhost, it is vscode's job to manage the window.

**Comment by isidorn** [comment]
@roblourens please note that we could not reproduce this with Electron 3 also. This only reproduces with Electron 5 and 6.
I do not know why port 9229 would be specified. We never set ports explicitly on the vscode side.

**Comment by alexandrudima** [comment]
@roblourens

> Also, the node debug adapter does not try to kill the extensionhost, it is vscode's job to manage the window.

VS Code does manage the window. In this case, VS Code is launched with some flags to indicate extension tests. VS Code will wait for the extension host process to exit and then it will close the window. 

But what appears to happen when pressing "Stop" while on a breakpoint is that the extension host process remains paused at the breakpoint. VS Code is not aware that the extension host has been debugged, and the debugger detached, and that the extension host is now permanently paused at a breakpoint.

This might be some weird node.js behavior (to leave the process paused at a breakpoint when the debugger detaches), but would it not make sense for the debugger (before detaching) to clean up a bit? I mean the debugger could remove its breakpoints and resume the process being debugged...

**Comment by roblourens** [comment]
The debug adapter doesn't kill the EH process, and if you're saying vscode waits for it to exit, I guess I'm not sure who is telling it to shut down. I can send Resume before detaching, that sounds harmless, although Node should take care of that for me.

Trying to repro again this morning, the only time I've seen it happen, I realized that the Extension Development Host was actually left open in the background after I clicked Stop. When I close that manually, the EH process goes down too. Is anyone else seeing that?

**Comment by roblourens** [comment]
Oops

![image](https://user-images.githubusercontent.com/323878/69443933-347f2700-0d04-11ea-807d-a25c3f55b0a5.png)

Hey @connor4312 could you try reproing this too?

**Comment by roblourens** [comment]
I think this is the line that should kill the window? https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/contrib/debug/browser/extensionHostDebugService.ts#L61

**Comment by roblourens** [comment]
edit, I deleted some comments because I realized that I misunderstood the issue. Anyway, I was able to repro this just using chrome devtools.

I can file an electron bug if I can reduce it, but I also tried to repro this just using node, and using normal EH debugging (not tests) and couldn't. Do you have any idea why it would be different when running tests?

**Comment by isidorn** [comment]
@roblourens note that the Test launch configuration has a lot of env variables being set.
It might be that one of those somehow breaks the flow
```
"env": {
                "MOCHA_grep": "UDF Malformed", // RegExp of tests to run (empty for all)
                "MOCHA_invert": "0", // Invert the RegExp
                "AZCODE_ARM_IGNORE_BUNDLE": "1",
                "MOCHA_enableTimeouts": "0", // Disable time-outs
                "DEBUGTELEMETRY": "1", // 1=quiet; verbose=see telemetry in console; 0=send telemetry
                "NODE_DEBUG": "",
                "DISABLE_LANGUAGE_SERVER_TESTS": "0",
                "DISABLE_SLOW_TESTS": "0"
            }
```

**Comment by roblourens** [comment]
I've managed to repro this with normal EH debugging in our vscode-pwa extension several times. @deepak1556 said that he didn't see the extension development host window getting whatever message it should get to shut down. I was expecting to see `ExtensionHostDebugChannelClient#close` called when I click the stop button but I don't see that, so I guess I don't know how shutting down the window is supposed to work at all. Can @isidorn or @weinand explain or help me figure out where to set a breakpoint?

**Comment by weinand** [comment]
Clicking the debugger's Stop button doesn't do anything special: it just terminates the extension host node.js process via the debug adapter.

VS Code should pick this up as a dying EH process [here](https://github.com/microsoft/vscode/blob/bee5357bedae47c11a631eacaa59a05e4dc4508c/src/vs/workbench/services/extensions/electron-browser/extensionHost.ts#L218).

And then it should not restart the EH process (because it is in debug mode) and should "exit" by closing the EH's window.

**Comment by roblourens** [comment]
Sorry, I insisted that the debug adapter doesn't kill the EH, and it totally is supposed to in EH debugging, different from normal "attach" configs.

Now node is sometimes returning `null` for `process.pid` right when it starts up. I added a workaround for that and it seems to be working.

**Comment by isidorn** [comment]
@roblourens Since you added a workaround, shouldn't this be closed?

**Comment by StephenWeatherford** [comment]
It does indeed seem to be fixed with 1.41.0-insider (user setup) 
Thanks!

**Comment by weinand** [comment]
Since @StephenWeatherford confirmed that this is now working for him and since I could no longer reproduce the issue, I'm adding the Verified tag.
